src/vi/mimihentai/src/eu/kanade/tachiyomi/extension/vi/mimihentai/MiMiHentai.kt

```diff
@@ -139,35 +143,87 @@ class MiMiHentai : HttpSource() {
                 when (filters) {
                     is SortByList ->
                         {
                             val sort = getSortByList()[filters.state]
                             addQueryParameter("sort", sort.id)
                         }
-
+                    is GenreList -> {
+                        filters.state.filter { it.state }
+                            .let { list ->
+                                if (list.isNotEmpty()) { list.forEach { genre -> addQueryParameter("genre", genre.id) } }
+                            }
+                    }
+                    is TextField -> setQueryParameter(filters.key, filters.state)
                     else -> {}
                 }
             }
         }.build()
         return GET(url, headers)
     }
 
+    private fun genresRequest(): Request = GET("$apiUrl/genres", headers)
+
+    private fun parseGenres(response: Response): List<Pair<Long, String>> {
+        return response.parseAs<List<Genres>>().map { Pair(it.id, it.name) }
+    }
+
+    private var fetchGenresAttempts: Int = 0
+    private fun fetchGenres() {
+        if (fetchGenresAttempts >= 3 || genreList.isEmpty()) {
+            launchIO {
+                try {
+                    client.newCall(genresRequest()).await()
+                        .use { parseGenres(it) }
+                        .takeIf { it.isNotEmpty() }
+                        ?.also { genreList = it }
+                } catch (_: Exception) {
+                } finally {
+                    fetchGenresAttempts++
+                }
+            }
+        }
+    }
+
+    private fun launchIO(block: suspend () -> Unit) = GlobalScope.launch(Dispatchers.IO) { block() }
+
+    private var genreList: List<Pair<Long, String>> = emptyList()
+
+    private class GenreList(name: String, pairs: List<Pair<Long, String>>) : GenresFilter(name, pairs)
+    private open class GenresFilter(title: String, pairs: List<Pair<Long, String>>) :
+        Filter.Group<GenreCheckBox>(title, pairs.map { GenreCheckBox(it.second, it.first.toString()) })
+    class GenreCheckBox(name: String, val id: String = name) : Filter.CheckBox(name)
+
     private class SortByList(sort: Array<SortBy>) : Filter.Select<SortBy>("Sắp xếp", sort)
     private class SortBy(name: String, val id: String) : Filter.CheckBox(name) {
         override fun toString(): String {
             return name
         }
     }
 
-    override fun getFilterList() = FilterList(
-        SortByList(getSortByList()),
-    )
+    private class TextField(name: String, val key: String) : Filter.Text(name)
+    override fun getFilterList(): FilterList {
+        fetchGenres()
+        return FilterList(
+            SortByList(getSortByList()),
+            TextField("Tác giả", "author"),
+            TextField("Parody", "parody"),
+            TextField("Nhân vật", "character"),
+            if (genreList.isEmpty()) {
+                Filter.Header("Thể loại")
```

Comment:
Usually this header should say "Tap 'Reset' to try to load the genres" so it's more helpful.